src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -7099,7 +7099,7 @@ private static bool HasImplicitTypeArguments(BoundNode node)
                 return true;
             }
 
-            if (node is BoundForEachStatement { EnumeratorInfoOpt: { GetEnumeratorInfo: { Method: { TypeArgumentsWithAnnotations: { IsEmpty: false } } } } })
+            if (node is BoundForEachStatement { EnumeratorInfoOpt: { GetEnumeratorInfo: { Method: var getEnumeratorMethod } } } && getEnumeratorMethod.GetMemberArityIncludingExtension() > 0)
```

Comment:
The call site of this method has already tested `GetMemberArityIncludingExtension() > 0`, checks of this kind within this method should probably be deleted, and the method could perhaps be moved to local function.